dali/kernels/shape.h

```diff
@@ -0,0 +1,299 @@
+  // TensorShape(const TensorShape<other_dim> &other) {
+  //   *this = other;
+  // }
+
+  // template <int other_dim>
+  // TensorShape &operator=(const TensorShape<other_dim> &other) {
+  //   data.resize(other.size());
+  //   for (int i = 0; i < size(); i++) {
+  //     data[i] = other[i];
+  //   }
+  //   return *this;
+  // }
+
+  // TensorShape &operator=(TensorShape<-1> &&other) {
+  //   data = std::move(other.data);
+  //   return *this;
+  // }
+
+  // We allow only explicit conversion from dynamic to static dim
+  // template <int other_dim>
+  // explicit operator TensorShape<other_dim>() const;
+  template <int other_dim>
+  TensorShape<other_dim> to_static() const;
+
+  int64_t &operator[](int d) { return shape[d]; }
+  const int64_t &operator[](int d) const { return shape[d]; }
+
+  std::vector<int64_t> shape;
+  int size() const { return shape.size(); }
+};
+
+template <int dim_>
+struct TensorShape {
+  TensorShape(const std::array<int64_t, dim_> &s) : shape(s) {
+    std::cout << "const std::array<int64_t, dim_> &s" << std::endl;
+  }
+  TensorShape(std::array<int64_t, dim_> &&s) : shape(std::move(s)) {
+    std::cout << "std::array<int64_t, dim_> &&s" << std::endl;
+  }
+
+  // We just want zero-initialization
+  TensorShape() : shape{} {}
+  // We allow only explicit operations on TensorShape static dim
+  TensorShape(const TensorShape &) = default;
+  TensorShape(TensorShape &&) = default;
+  TensorShape &operator=(const TensorShape &other) = default;
+  TensorShape &operator=(TensorShape &&other) = default;
+
+  template <typename... Ts>
+  TensorShape(int64_t i0, Ts... s) : shape{i0, int64_t{s}...} {
+    static_assert(sizeof...(Ts) == dim_ - 1, "Number of shapes passed must match dim_");
+    std::cout << "TensorShape(int64_t i0, Ts... s)" << std::endl;
+    // TODO static_assert for Ts == int64_t
+  }
+
+  int64_t &operator[](int d) { return shape[d]; }
+  const int64_t &operator[](int d) const { return shape[d]; }
+
+  std::array<int64_t, dim_> shape;
+  constexpr int size() const { return dim_; }
+};
+
+// template<int other_dim>
+// TensorShape<DynamicTensorShape>::operator TensorShape<other_dim>() const {
+//   std::cout << "operator TensorShape<other_dim>()" << std::endl;
+//   assert(size() == other_dim);
+//   TensorShape<other_dim> shape;
+//   for (int i = 0; i < other_dim; i++) {
+//     shape[i] = (*this)[i];
+//   }
+//   return shape;
+// }
+
+template <int other_dim>
+TensorShape<other_dim> TensorShape<DynamicTensorShape>::to_static() const {
+  std::cout << "operator TensorShape<other_dim>()" << std::endl;
+  assert(size() == other_dim);
+  TensorShape<other_dim> shape;
+  for (int i = 0; i < other_dim; i++) {
+    shape[i] = (*this)[i];
+  }
+  return shape;
+}
+
+// TODO should we be able to only output a tensor of sample_dim dim for all cases of TensorListShape
+template <int sample_dim_>
+struct TensorListShape {
+  TensorListShape(std::vector<TensorShape<sample_dim_>> sample_shapes)
```

Comment:
Have you looked at my PR? IMO we should take the number of dimensions of the first sample and then enforce that the sizes match for other shapes. If we pass dynamic shapes here, we should validate that they have appropriate number of dimensions (matching the first one or the static sample_dim_, if not dynamic). Unlike tensor shape, this is probably going to be a frequent case, that we want to convert a fully dynamic shape into a semi-static one.

Reply:
I'm rather thinking, that if you are creating the `TLV` or `TV`, most of the time you already have the shape of the data validated, etc. At least in case of static dim shapes. Didn't touch the dynamic ones yet, will think about how to cover that, maybe some utility functions will be able to cover different cases.